### PR TITLE
[CELEBORN-659][SPARK][TEST] Refine RssShuffleWriterSuiteJ

### PR DESCRIPTION
**Pull request by cfmcgrady**
### What changes were proposed in this pull request?

1. renamed `RssShuffleWriterSuiteJ` to `CelebornShuffleWriterSuiteBase`, which now serves as an abstract base class.
2. two new classes, `HashBasedShuffleWriterSuiteJ` and `SortBasedShuffleWriterSuiteJ`, have been added. These classes extend `CelebornShuffleWriterSuiteBase` and provide suites for testing hash-based and sort-based shuffle writers.

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
Also change for Spark2?

File: client-spark/spark-3/src/test/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriterSuiteJ.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+      RssShuffleHandle handle,
+      TaskContext context,
+      CelebornConf conf,
+      ShuffleClient client,
+      ShuffleWriteMetricsReporter metrics)
+      throws IOException {
+    return new SortBasedShuffleWriter<Integer, String, String>(
+        handle.dependency(),
+        handle.appUniqueId(),
+        handle.numMappers(),
```
Comment (pan3793) [inline client-spark/spark-3/src/test/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriterSuiteJ.java:41]: Maybe we can change the SortBasedShuffleWriter constructor and pass handle directly?

**Comment by cfmcgrady** [comment]
> Also change for Spark2?

Spark 2 already includes the suite for the sort-based shuffle writer. In order to maintain consistency with Spark 3, the logic of `CelebornShuffleWriterSuiteBase.java` from Spark 2 is simplified and aligned with Spark 3.

**Comment by cfmcgrady** [comment]
```diff
diff --git client-spark/spark-3/src/test/java/org/apache/spark/shuffle/celeborn/CelebornShuffleWriterSuiteBase.java client-spark/spark-2/src/test/java/org/apache/spark/shuffle/celeborn/CelebornShuffleWriterSuiteBase.java
index 6995ae66..39a039ef 100644
--- client-spark/spark-3/src/test/java/org/apache/spark/shuffle/celeborn/CelebornShuffleWriterSuiteBase.java
+++ client-spark/spark-2/src/test/java/org/apache/spark/shuffle/celeborn/CelebornShuffleWriterSuiteBase.java
@@ -50,11 +50,11 @@ import org.apache.spark.memory.UnifiedMemoryManager;
 import org.apache.spark.scheduler.MapStatus;
 import org.apache.spark.serializer.KryoSerializer;
 import org.apache.spark.serializer.Serializer;
-import org.apache.spark.shuffle.ShuffleWriteMetricsReporter;
 import org.apache.spark.shuffle.ShuffleWriter;
 import org.apache.spark.sql.catalyst.InternalRow;
 import org.apache.spark.sql.catalyst.expressions.UnsafeProjection;
 import org.apache.spark.sql.catalyst.expressions.UnsafeRow;
+import org.apache.spark.sql.execution.PartitionIdPassthrough;
 import org.apache.spark.sql.execution.UnsafeRowSerializer;
 import org.apache.spark.sql.types.DataType;
 import org.apache.spark.sql.types.IntegerType$;
@@ -79,7 +79,6 @@ import org.apache.celeborn.common.CelebornConf;
 import org.apache.celeborn.common.identity.UserIdentifier;
 import org.apache.celeborn.common.util.JavaUtils;
 import org.apache.celeborn.common.util.Utils;
-import org.apache.celeborn.reflect.DynConstructors;
 
 public abstract class CelebornShuffleWriterSuiteBase {
 
@@ -97,12 +96,13 @@ public abstract class CelebornShuffleWriterSuiteBase {
   private final UserIdentifier userIdentifier = new UserIdentifier("mock", "mock");
 
   private final int numMaps = 10;
-  private final int numPartitions = 10;
+  protected final int numPartitions = 10;
   private final SparkConf sparkConf = new SparkConf(false);
   private final BlockManagerId bmId = BlockManagerId.apply("execId", "host", 1, None$.empty());
-
   private final TaskMemoryManager taskMemoryManager =
       new TaskMemoryManager(UnifiedMemoryManager.apply(sparkConf, 1), 0);
+  private final MapStatus mapStatus =
+      SparkUtils.createMapStatus(bmId, new long[numPartitions], new long[numPartitions]);
 
   @Mock(answer = Answers.RETURNS_SMART_NULLS)
   private TaskContext taskContext = null;
@@ -120,6 +120,8 @@ public abstract class CelebornShuffleWriterSuiteBase {
 
   private static File tempDir = null;
 
+  public CelebornShuffleWriterSuiteBase() throws IOException {}
+
   @BeforeClass
   public static void beforeAll() {
     tempDir = Utils.createTempDir(System.getProperty("java.io.tmpdir"), "celeborn_test");
@@ -216,17 +218,8 @@ public abstract class CelebornShuffleWriterSuiteBase {
       throws Exception {
     final boolean useUnsafe = serializer instanceof UnsafeRowSerializer;
 
-    DynConstructors.Ctor<Partitioner> partitionIdPassthroughCtor =
-        DynConstructors.builder()
-            // for Spark 3.3 and previous
-            .impl("org.apache.spark.sql.execution.PartitionIdPassthrough", int.class)
-            // for Spark 3.4
-            .impl("org.apache.spark.PartitionIdPassthrough", int.class)
-            .build();
     final Partitioner partitioner =
-        useUnsafe
-            ? partitionIdPassthroughCtor.newInstance(numPartitions)
-            : new HashPartitioner(numPartitions);
+        useUnsafe ? new PartitionIdPassthrough(numPartitions) : new HashPartitioner(numPartitions);
     Mockito.doReturn(partitioner).when(dependency).partitioner();
     Mockito.doReturn(serializer).when(dependency).serializer();
 
@@ -237,7 +230,7 @@ public abstract class CelebornShuffleWriterSuiteBase {
     ((DummyShuffleClient) client).initReducePartitionMap(shuffleId, numPartitions, 1);
 
     final ShuffleWriter<Integer, String> writer =
-        createShuffleWriter(handle, taskContext, conf, client, metrics.shuffleWriteMetrics());
+        createShuffleWriter(handle, taskContext, conf, client);
 
     if (writer instanceof SortBasedShuffleWriter) {
       assertEquals(useUnsafe, ((SortBasedShuffleWriter) writer).canUseFastWrite());
@@ -360,10 +353,6 @@ public abstract class CelebornShuffleWriterSuiteBase {
   }
 
   protected abstract ShuffleWriter<Integer, String> createShuffleWriter(
-      RssShuffleHandle handle,
-      TaskContext context,
-      CelebornConf conf,
-      ShuffleClient client,
-      ShuffleWriteMetricsReporter metrics)
+      RssShuffleHandle handle, TaskContext context, CelebornConf conf, ShuffleClient client)
       throws IOException;
 }
```

File: client-spark/spark-3/src/test/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriterSuiteJ.java (hunk 1 above)
Comment (cfmcgrady) [reply client-spark/spark-3/src/test/java/org/apache/spark/shuffle/celeborn/SortBasedShuffleWriterSuiteJ.java]: introduce a new constructor

```java  
public SortBasedShuffleWriter(
      RssShuffleHandle<K, V, C> handle,
      TaskContext taskContext,
      CelebornConf conf,
      ShuffleClient client,
      ShuffleWriteMetricsReporter metrics,
      ExecutorService executorService)
      throws IOException
```

File: client-spark/spark-2/src/test/java/org/apache/spark/shuffle/celeborn/HashBasedShuffleWriterSuiteJ.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+public class HashBasedShuffleWriterSuiteJ extends CelebornShuffleWriterSuiteBase {
+
+  public HashBasedShuffleWriterSuiteJ() throws IOException {}
+
+  @Override
+  protected ShuffleWriter<Integer, String> createShuffleWriter(
+      RssShuffleHandle handle, TaskContext context, CelebornConf conf, ShuffleClient client)
+      throws IOException {
+    return new HashBasedShuffleWriter<Integer, String, String>(
+        handle, 0, context, conf, client, SendBufferPool.get(1));
```
Comment (waitinfuture) [inline client-spark/spark-2/src/test/java/org/apache/spark/shuffle/celeborn/HashBasedShuffleWriterSuiteJ.java:37]: Passing 0 as mapId is a little bit strange here, although originally the mapId is also final 0. Shall we add a comment here?

**Review by waitinfuture — APPROVED** [review]
LGMT except a nit, thanks for the refactoring!

**Review by AngersZhuuuu — APPROVED** [review]
Nice change

File: client-spark/spark-2/src/test/java/org/apache/spark/shuffle/celeborn/HashBasedShuffleWriterSuiteJ.java (hunk 2 above)
Comment (cfmcgrady) [reply client-spark/spark-2/src/test/java/org/apache/spark/shuffle/celeborn/HashBasedShuffleWriterSuiteJ.java]: named the input argument and added a comment.

**Pull request closed without merging**